src/topology/constructions.lean

```diff
@@ -154,6 +154,45 @@ lemma continuous.prod_map {f : γ → α} {g : δ → β} (hf : continuous f) (h
   continuous (λ x : γ × δ, (f x.1, g x.2)) :=
 (hf.comp continuous_fst).prod_mk (hg.comp continuous_snd)
 
+/-- A version of `continuous_inf_dom_left` for binary functions -/
+lemma continuous_inf_dom_left₂ {α β γ} {f : α → β → γ}
+  {ta1 ta2 : topological_space α} {tb1 tb2 : topological_space β} {tc1 : topological_space γ}
+  (h : by haveI := ta1; haveI := tb1; exact continuous (λ p : α × β, f p.1 p.2)) :
+  by haveI := ta1 ⊓ ta2; haveI := tb1 ⊓ tb2; exact continuous (λ p : α × β, f p.1 p.2) :=
+begin
+  have ha := @continuous_inf_dom_left _ _ id ta1 ta2 ta1 (@continuous_id _ (id _)),
+  have hb := @continuous_inf_dom_left _ _ id tb1 tb2 tb1 (@continuous_id _ (id _)),
+  have h_continuous_id := @continuous.prod_map _ _ _ _ ta1 tb1 (ta1 ⊓ ta2) (tb1 ⊓ tb2) _ _ ha hb,
+  exact @continuous.comp _ _ _ (id _) (id _) _ _ _ h h_continuous_id,
+end
+
+/-- A version of `continuous_inf_dom_right` for binary functions -/
+lemma continuous_inf_dom_right₂ {α β γ} {f : α → β → γ}
+  {ta1 ta2 : topological_space α} {tb1 tb2 : topological_space β} {tc1 : topological_space γ}
+  (h : by haveI := ta2; haveI := tb2; exact continuous (λ p : α × β, f p.1 p.2)) :
+  by haveI := ta1 ⊓ ta2; haveI := tb1 ⊓ tb2; exact continuous (λ p : α × β, f p.1 p.2) :=
+begin
+  have ha := @continuous_inf_dom_right _ _ id ta1 ta2 ta2 (@continuous_id _ (id _)),
+  have hb := @continuous_inf_dom_right _ _ id tb1 tb2 tb2 (@continuous_id _ (id _)),
+  have h_continuous_id := @continuous.prod_map _ _ _ _ ta2 tb2 (ta1 ⊓ ta2) (tb1 ⊓ tb2) _ _ ha hb,
+  exact @continuous.comp _ _ _ (id _) (id _) _ _ _ h h_continuous_id,
+end
+
+/-- A version of `continuous_Inf_dom` for binary functions -/
+lemma continuous_Inf_dom₂ {α β γ} {f : α → β → γ}
+  {tas : set (topological_space α)} {tbs : set (topological_space β)}
+  {ta : topological_space α} {tb : topological_space β} {tc : topological_space γ}
+  (ha : ta ∈ tas) (hb : tb ∈ tbs)
+  (hf : continuous (λ p : α × β, f p.1 p.2)):
+  by haveI := Inf tas; haveI := Inf tbs; exact @continuous _ _ _ tc (λ p : α × β, f p.1 p.2) :=
+begin
+  let t : topological_space (α × β) := prod.topological_space,
+  have ha := continuous_Inf_dom ha continuous_id,
+  have hb := continuous_Inf_dom hb continuous_id,
+  have h_continuous_id := @continuous.prod_map _ _ _ _ ta tb (Inf tas) (Inf tbs) _ _ ha hb,
+  exact @continuous.comp _ _ _ (id _) (id _) _ _ _ hf h_continuous_id,
+end
```

Comment:
These proofs are awful - is there an easy way to make the underscores go away?

Reply:
Actually, I was messing with this for a bit and I just think there's too many unification issues, because you're not only overriding TC (`id _` is a crazy hack for that, btw) but also you're asking lean to unify `f ∘ ?m_1 = f` with `f ∘ id`, which I think it's reluctant to do because there's many other possible solutions (which may be defeq too)

some toying I did with this:

```lean
lemma continuous.comp' {α β γ} [topological_space α]
  (hb : topological_space β) [hg : topological_space γ] {g : β → γ} (f : α → β) :
  continuous g → continuous f → continuous (g ∘ f) := continuous.comp

set_option pp.implicit true

/-- A version of `continuous_inf_dom_left` for binary functions -/
lemma continuous_inf_dom_left₂ {α β γ} {f : α → β → γ}
  {ta1 ta2 : topological_space α} {tb1 tb2 : topological_space β} {tc1 : topological_space γ}
  (h : by haveI := ta1; haveI := tb1; exact continuous (λ p : α × β, f p.1 p.2)) :
  by haveI := ta1 ⊓ ta2; haveI := tb1 ⊓ tb2; exact continuous (λ p : α × β, f p.1 p.2) :=
begin
  refine continuous.comp' (@@prod.topological_space (ta1 ⊓ ta2) (tb1 ⊓ tb2)) (λ p : α × β, (p.1, p.2)) _ _,

```

Reply:
It seems the way to fix this is to replace the `[topological_space X]` arguments with `{tx : topology_space X}`, like I think @RemyDegenne did with lots of the probability theory stuff. That blows up the size of this PR considerably though, so makes sense to leave to a follow-up.